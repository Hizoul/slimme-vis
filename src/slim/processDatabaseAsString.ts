
export interface IClass {
  label: string
  values: number
}

export interface Database {[index: string]: number[][] | undefined}

const processDatabaseAsString = (databaseContent: string) => {
  const classes: Database = {}
  for (const line of databaseContent.split("\n")) {
    const values = line.split(" ").map((value) => {
      if (value.length > 0) {
        return Number(value)
      }
    }).filter((val) => val != null)
    const classKey = String(values[0])
    if (classes[classKey] == null) {
      classes[classKey] = []
    }
    classes[classKey].push(values.splice(1))
  }
  return classes
}

export default processDatabaseAsString
