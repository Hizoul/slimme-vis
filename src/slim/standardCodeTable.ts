import isEqual from "../util/isEqual"
import { Database } from "./processDatabaseAsString"

export interface ICodeTable {
  left: any
  right: any
}

export interface CodeTable {[index: string]: number}
export type ICodeTableEntry = number[][]
export type SortedCodeTable = ICodeTableEntry[]

const standardCodeTable = (dataset: Database) => {
  const itemSets: CodeTable = {}
  for (const key of Object.keys(dataset)) {
    for (const itemSet of dataset[key]) {
      for (const value of itemSet) {
        const itemKey = String(value)
        if (itemSets[itemKey] == null) {
          itemSets[itemKey] = 0
        }
        itemSets[itemKey]++
      }
    }
  }
  return itemSets
}

const updateFrequencies = (dataset: Database, codeTable: SortedCodeTable) => {
  const clonedDb: Database = JSON.parse(JSON.stringify(dataset))
  for (const entry of codeTable) {
    entry[1][0] = 0
    for (const classKey of Object.keys(clonedDb)) {
      for (const transactionIndex in clonedDb[classKey]) {
        const transaction = clonedDb[classKey][transactionIndex]
        if (subset(entry[0], transaction).length === entry[0].length) {
          entry[1][0]++
          clonedDb[classKey][transactionIndex] = transaction.filter((value) => entry[0].indexOf(value) === -1)
        }
      }
    }
  }
  return standardCoverOrder(codeTable)
}

const supportOfItem = (itemSet: number[], dataset: Database) => {
  let support = 0
  for (const key of Object.keys(dataset)) {
    for (const set of dataset[key]) {
      if (subset(itemSet, set).length === itemSet.length) {
        support++
      }
    }
  }
  return support
}

const sortedStandardCodeTable = (dataset: Database) => {
  const itemSets: CodeTable = standardCodeTable(dataset)
  const sortedCodeTable: SortedCodeTable = []
  for (const key of Object.keys(itemSets)) {
    sortedCodeTable.push([[Number(key)], [itemSets[key]], [supportOfItem([itemSets[key]], dataset)]])
  }
  return standardCoverOrder(sortedCodeTable)
}

// first descending on cardinality (amount of values), second descending on support (here called frequency), and third lexicographically
const standardCoverOrder = (codeTable: SortedCodeTable) => {
  codeTable.sort((itemA, itemB) => {
    if (itemA[0].length < itemB[0].length) {
      return 1
    }
    if (itemA[0].length > itemB[0].length) {
      return -1
    }
    if (itemA[2][0] < itemB[2][0]) {
      return 1
    }
    if (itemA[2][0] > itemB[2][0]) {
      return -1
    }
    return -1 * JSON.stringify(itemA[0]).localeCompare(JSON.stringify(itemB[0]))
  })
  return codeTable
}

const subset = (arrA: any[], arrB: any[]) => {
  const arrToUse = arrA.length > arrB.length ? arrB : arrA
  const arrToCompare = arrA.length > arrB.length ? arrA : arrB
  const result: any[] = []
  arrToUse.map((value) => {
    if (arrToCompare.indexOf(value) !== -1) {
      result.push(value)
    }
  })
  return result
}

export default sortedStandardCodeTable
export {
  standardCoverOrder, subset, updateFrequencies, supportOfItem
}
