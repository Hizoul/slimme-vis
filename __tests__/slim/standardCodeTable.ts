import { sha512 } from "js-sha512"
import { resolve } from "path"
import processLocalDb from "../../src/node/local"
import processDatabaseAsString from "../../src/slim/processDatabaseAsString"
import standardCodeTable, { standardCoverOrder } from "../../src/slim/standardCodeTable"

const database = `
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2
0 1
0 2
`
describe("ProcessDB", () => {
  it("should process local file", () => {
    const dataset = processDatabaseAsString(database)
    expect(
      standardCodeTable(dataset)
    ).toMatchSnapshot("standard code table of mushroomset")
    expect(
      sha512(JSON.stringify(dataset))
    ).toMatchSnapshot("before sort")
  })
  it("cover sort algo", () => {
    expect(standardCoverOrder([
      [[100, 102], [100]],
      [[100], [100]]
    ])).toMatchSnapshot("result of sort")
    expect(standardCoverOrder([
      [[100], [100]],
      [[100, 102], [100]]
    ])).toMatchSnapshot("result of sort")
  })
})
