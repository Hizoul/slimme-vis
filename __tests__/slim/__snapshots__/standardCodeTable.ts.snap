// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessDB cover sort algo: result of sort 1`] = `
Array [
  Array [
    Array [
      100,
      102,
    ],
    Array [
      100,
    ],
  ],
  Array [
    Array [
      100,
    ],
    Array [
      100,
    ],
  ],
]
`;

exports[`ProcessDB cover sort algo: result of sort 2`] = `
Array [
  Array [
    Array [
      100,
      102,
    ],
    Array [
      100,
    ],
  ],
  Array [
    Array [
      100,
    ],
    Array [
      100,
    ],
  ],
]
`;

exports[`ProcessDB should process local file: before sort 1`] = `"caefdd1c029de137073f2a3f3f7dcd830b1be321c6bd4c6d20b1fa34a1b2dccf6db11d8d9e0448916a3fa41d1512ef1755b596fe273c9bdf2f6a521a58c92db0"`;

exports[`ProcessDB should process local file: standard code table of mushroomset 1`] = `
Array [
  Array [
    Array [
      2,
    ],
    Array [
      7,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      7,
    ],
  ],
  Array [
    Array [
      3,
    ],
    Array [
      5,
    ],
  ],
]
`;
